packages/pigeon/lib/cpp_generator.dart

```diff
@@ -0,0 +1,794 @@
+          });
+        }
+        indent.write('default:');
+        indent.writeScoped('', '', () {
+          indent.writeln(
+              'return flutter::StandardCodecSerializer::ReadValueOfType(type, stream);');
+        });
+      });
+    });
+    indent.writeln('');
+    indent.write(
+        'void $codecName::WriteValue(const flutter::EncodableValue& value, flutter::ByteStreamWriter* stream) const');
+    indent.writeScoped('{', '}', () {
+      indent.write(
+          'if (const flutter::CustomEncodableValue* custom_value = std::get_if<flutter::CustomEncodableValue>(&value))');
+      indent.scoped('{', '}', () {
+        for (final EnumeratedClass customClass in getCodecClasses(api, root)) {
+          indent.write(
+              'if (custom_value->type() == typeid(${customClass.name}))');
+          indent.scoped('{', '}', () {
+            indent.writeln('stream->WriteByte(${customClass.enumeration});');
+            indent.writeln(
+                'WriteValue(std::any_cast<${customClass.name}>(*custom_value).ToEncodableMap(), stream);');
+            indent.writeln('return;');
+          });
+        }
+      });
+      indent.writeln(
+          'flutter::StandardCodecSerializer::WriteValue(value, stream);');
+    });
+  }
+}
+
+void _writeErrorOr(Indent indent) {
+  indent.format('''
+class FlutterError {
+ public:
+\tFlutterError();
+\tFlutterError(std::string arg_code)
+\t\t: code(arg_code) {};
+\tFlutterError(std::string arg_code, std::string arg_message)
+\t\t: code(arg_code), message(arg_message) {};
+\tFlutterError(std::string arg_code, std::string arg_message, std::string arg_details)
+\t\t: code(arg_code), message(arg_message), details(arg_details) {};
+\tstd::string code;
+\tstd::string message;
+\tstd::string details;
+};
+template<class T> class ErrorOr {
+\tstd::variant<T, FlutterError> v;
+\tbool ok = true;
+ public:
+\tErrorOr() { new(&v) T(); }
+\tErrorOr(const T& rhs) { new(&v) T(rhs); }
+\tErrorOr(const FlutterError& rhs) : ok(false) {
+\t\tnew(&v) FlutterError(rhs);
+\t}
+\tbool hasError() const { return !ok; }
+\tconst T& value() const { return std::get<T>(v); };
+\tconst FlutterError& error() const { return std::get<FlutterError>(v); };
+};
+''');
+}
+
+void _writeHostApiHeader(Indent indent, Api api) {
+  assert(api.location == ApiLocation.host);
+
+  indent.writeln(
+      '/* Generated class from Pigeon that represents a handler of messages from Flutter. */');
+  indent.write('class ${api.name} ');
+  indent.scoped('{', '};', () {
+    indent.scoped(' public:', '', () {
+      indent.writeln('${api.name}(const ${api.name}&) = delete;');
+      indent.writeln('${api.name}& operator=(const ${api.name}&) = delete;');
+      indent.writeln('virtual ~${api.name}() { };');
+      for (final Method method in api.methods) {
+        final String returnTypeName = method.returnType.isVoid
+            ? 'std::optional<FlutterError>'
```

Comment:
I'm wondering about the ergonomics of this; it's essentially a double-nested `variant` to express three states. Should we instead have a `NullableErrorOr` that can directly express the three states? It seems (naively; I haven't tried implementing an API using this yet, so maybe you've found that it's not the case) that it would be easier to author implementations that way.

Reply:
I wanted to leverage what was already implemented by std as much as possible, so this seemed like the best approach.